src/types/redis_bloomfilter.cc

```diff
@@ -0,0 +1,378 @@
+    std::string sub_key, value;
+    InternalKey(bf_key, std::to_string(index), bf_metadata.version, storage_->IsSlotIdEncoded()).Encode(&sub_key);
+    rocksdb::Status s =
+        storage_->Get(rocksdb::ReadOptions(), sub_key, &value);  // todo: use bf_key in InternalKey maybe error
+    if (!s.ok() && !s.IsNotFound()) return s;
+    uint32_t byte_index = (x / 8) % kBloomFilterSegmentBytes;
+    if (byte_index >= value.size()) {  // expand the bloomfilter
+      size_t expand_size = 0;
+      if (byte_index >= value.size() * 2) {
+        expand_size = byte_index - value.size() + 1;
+      } else if (value.size() * 2 > kBloomFilterSegmentBytes) {
+        expand_size = kBloomFilterSegmentBytes - value.size();
+      } else {
+        expand_size = value.size();
+      }
+      value.append(expand_size, 0);
+    }
+
+    if (!testBitSetBit(value, x, mode)) {
+      if (mode == ReadWriteMode::MODE_READ) {
+        ret = 0;
+        return rocksdb::Status::OK();
+      } else {
+        found_unset = 1;
+        batch->Put(sub_key, value);
+        storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+        // todo: if the indexs are same, the value need to be updated immediately, if have other ways to compl it
+      }
+    }
+  }
+  if (mode == ReadWriteMode::MODE_READ) {
+    ret = 1;
+  } else {
+    ret = found_unset;
+  }
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status BloomFilter::Reserve(const Slice &user_key, double error_rate, uint64_t capacity, uint16_t expansion,
+                                     uint16_t scaling) {
+  std::string ns_key;
+  AppendNamespacePrefix(user_key, &ns_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  SBChainMetadata sb_chain_metadata;
+  rocksdb::Status s = getSBChainMetadata(ns_key, &sb_chain_metadata);
+  if (!s.ok() && !s.IsNotFound()) return s;
+  if (!s.IsNotFound()) {
+    return rocksdb::Status::InvalidArgument("item exists");
+  }
+
+  return createSBChain(ns_key, error_rate, capacity, expansion, scaling, sb_chain_metadata);
+}
+
+rocksdb::Status BloomFilter::Add(const Slice &user_key, const Slice &item, int &ret) {
+  std::vector<int> tmp(1, 0);
+  rocksdb::Status s = MAdd(user_key, {item}, tmp);
+  ret = tmp[0];
+  return s;
+}
+
+rocksdb::Status BloomFilter::MAdd(const Slice &user_key, const std::vector<Slice> &items, std::vector<int> &rets) {
+  BFInsertOptions insert_options = {.capacity = kBFDefaultInitCapacity,
+                                    .error_rate = kBFDefaultErrorRate,
+                                    .autocreate = 1,
+                                    .expansion = kBFDefaultExpansion,
+                                    .scaling = 1};
+  return InsertCommon(user_key, items, &insert_options, rets);
+}
+
+rocksdb::Status BloomFilter::InsertCommon(const Slice &user_key, const std::vector<Slice> &items,
+                                          const BFInsertOptions *insert_options, std::vector<int> &rets) {
```

Comment:
```suggestion
                                          const BFInsertOptions& insert_options, std::vector<int>* rets) {
```

I think it's better to mark `insert_options` as const reference here.